### DISCUSSION TITLE
Rule to limit count of record by user in collection

### DISCUSSION THREAD

**Discussion by whateverneveranywhere** [pocketbase/pocketbase#2293] [Q&A] [answered]
Hey there!

Currently, I have a collection named Tasks, and each user can have multiple tasks.
So, when a user is creating a new Task, I'm checking the @request.auth.id to be matching the relational user that is in a row in the Tasks collection (table).

The case is that I want to only allow each user to have 2 tasks at a time. So, whenever the relational user record in each row in the Tasks collection is smaller or equal to 2, it's fine and is allowed to CREATE a new record. 

I couldn't find a workaround for this using custom rules in the dashboard app.
Am I even doing it correctly? 

Here's a simplified version of how my collection of Tasks looks for example:

```
[id: 1, task: 'Learn GO'],
[id: 2, task: 'Try Next JS 13'],
[id: 1, task: 'Go touch some grass'],
```
now I need a rule that will reject this record from getting created, because count of records in Tasks collection associated to the user with id of 1 is already at max, which is 2:
```
[id: 1, task: 'Read more books'],
```

Would appreciate any tips!

**Comment by ganigeorgiev** [comment, marked as answer]
There is a [`:length` modifier](https://pocketbase.io/docs/api-rules-and-filters/#length-modifier) but that will work only if the the user-task `relation` is in the `users` collection, aka. each user to have a multiple `relation` field named tasks.

But if the user-task `relation` field is in the `tasks` collection then what you are trying to do can't be done directly with just an API rule since we'll have to perform a special count subquery (_and it can become even more complicated if we have to specify additional conditions for the count_).

One slightly hacky workaround could be to create a helper "View collection" (_Admin UI > Create collection > Type View_) with the following as query (_you may have to change the column names to match with your `tasks` collection_):

```sql
-- the view collection could be named for example userTasksCount
SELECT
    tasks.user AS id,
    count(*) as total
FROM tasks
GROUP BY tasks.user
```

The above will create a read-only collection that will aggregate the number of tasks for each user (_the user id is used as id since it will be unique for each row_).

Then in your `tasks` collection, you can use the following rule:

```js
@collection.userTasksCount.id ?= user && 
@collection.userTasksCount.total ?<= 2
```
_Roughly translates to: at least one `userTasksCount` matches the current submitted user id and its total number of tasks (including the current created one) is less than or equal to 2. If the user tries to create 3 tasks it should fail with generic 400 error._


---

As always, another option could be also to perform manually the count query check in a [`OnRecordBeforeCreateRequest` Go hook](https://pocketbase.io/docs/event-hooks/#record-crud-api-hooks).

**Reply by Lex-2008** [reply]
Nice stuff! I also like how using views adds full SQL(ite) power to PocketBase API rules. Currently trying to implement multi-user [battleship][bs] game, using PocketBase with only views and API rules as a backend.

[bs]: https://en.wikipedia.org/wiki/Battleship_(game)

I wonder if it's worth mentioning somewhere in the docs that _create_ rule is checked _after_ the relevant record (tries to be) created in the collection? For example, note how in your answer above, to allow users create **two** tasks, in "create" rule you have to check `....userTasksCount.total ?<= 2`. My initial expectation was be that "create" rule is checked _before_ the record is created, and this rule would _allow a third_ record to be created.

**Reply by ganigeorgiev** [reply]
@Lex-2008 I'll consider mentioning it in the docs with the next update.
Just to clarify, internally the Create API rule is executed after a "dry submit" of the record (aka. partial save without the file uploads within a one-off transaction). If the rule resolves to true we persist the changes (including the uploaded files), otherwise - rollback the transaction and return 40x error.

**Comment by whateverneveranywhere** [comment]
Hi @ganigeorgiev,

Awesome suggestions, thank you so much!

Not sure if adding Tasks as a relation to the Users collection could be the best idea for my use case,
But the hacky workaround seems more flexible so far.
